go/genkit/session.go

```diff
@@ -0,0 +1,298 @@
+	// Schema for state variables
+	StateSchema *jsonschema.Schema `json:"stateschema,omitempty"`
+	// The messages for each thread
+	Threads map[string][]*ai.Message `json:"threads,omitempty"`
+}
+
+type Session struct {
+	// Genkit
+	Genkit *Genkit
+	// The session id
+	ID string
+	// The data for the session
+	SessionData SessionData
+	// Default state variable values
+	DefaultState map[string]any
+	// The store for the session, defaults to in-memory storage
+	Store SessionStore
+}
+
+// SessionOption configures params for the session
+type SessionOption func(s *Session) error
+
+// A session key
+type sessionKey string
+
+const (
+	currentSessionID    sessionKey = "currentSessionID"
+	currentSessionStore sessionKey = "currentSessionStore"
+)
+
+// NewSession creates a new session with the provided options.
+// If no store is provided, it defaults to in-memory storage.
+func NewSession(ctx context.Context, opts ...SessionOption) (session *Session, err error) {
+	s := &Session{}
+
+	for _, with := range opts {
+		err := with(s)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if s.Store == nil {
+		s.Store = &InMemSessionStore{}
+	}
+
+	if s.ID == "" {
+		s.ID = uuid.New().String()
+	}
+
+	if s.SessionData.Threads == nil {
+		s.SessionData.Threads = make(map[string][]*ai.Message)
+	}
+
+	// Only update state with defaults if not already set, eg. via WithSessionData
+	if s.SessionData.State == nil {
+		s.SessionData.State = s.DefaultState
+	}
+
+	// Initialize session in store
+	s.Store.Save(s.ID, s.SessionData)
+
+	return s, nil
+}
+
+// LoadSession loads sessiondata from store, and returns a session.
+func LoadSession(ctx context.Context, sessionId string, store SessionStore) (session *Session, err error) {
+	sessionData, err := store.Get(sessionId)
+	if err != nil {
+		return nil, err
+	}
+
+	session = &Session{
+		ID:          sessionId,
+		SessionData: sessionData,
+		Store:       store,
+	}
+
+	return session, nil
+}
+
+// UpdateState takes any data and sets it as state in the store.
+func (s *Session) UpdateState(state any) error {
+	var err error
+
+	// Ensure sessionData is up to date
+	sessionData, err := s.Store.Get(s.ID)
+	if err != nil {
+		return err
+	}
+	s.SessionData = sessionData
+
+	s.SessionData.StateSchema, s.SessionData.State, err = getSchemaAndDefaults(state)
+	if err != nil {
+		return err
+	}
+
+	return s.Store.Save(s.ID, s.SessionData)
+}
+
+// UpdateMessages takes a threadName and a slice of messages.
+func (s *Session) UpdateMessages(threadName string, messages []*ai.Message) error {
+	// Ensure sessionData is up to date
+	sessionData, err := s.Store.Get(s.ID)
+	if err != nil {
+		return err
+	}
+	s.SessionData = sessionData
+
+	s.SessionData.Threads[threadName] = messages
+	return s.Store.Save(s.ID, s.SessionData)
+}
+
+// Set current session ID in context.
+func (s *Session) SetContext(ctx context.Context) context.Context {
+	ctx = context.WithValue(ctx, currentSessionID, s.ID)
```

Comment:
We have context helpers in https://github.com/firebase/genkit/blob/72cc19ab00eb1d01c3c4236d4217d862762293c8/go/internal/base/context_key.go that might be useful here.